src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+        targetStatusList = RequestStatus.REQUEST_STATE_TRANSITION.get(wflow['RequestStatus'], [])
+        self.logger.info("targetStatusList: %s", targetStatusList)
+        return wflow
+
+    def getMSOutputTransferInfo(self, wflow):
+        """
+        Fetches the transfer information from the MSOutput REST interface for
+        the given workflow.
+        :param  wflow:   A MSRuleCleaner workflow representation
+        :return wflow:
+        """
+        headers = {'Accept': 'application/json'}
+        params = {}
+        url = '%s/data/info?request=%s' % (self.msConfig['msOutputUrl'],
+                                           wflow['RequestName'])
+        mgr = RequestHandler()
+        res = mgr.getdata(url, params=params, headers=headers, ckey=ckey(), cert=cert())
+        data = json.loads(res)['result'][0]
+        TransferInfo = data['transferDoc']
+        if TransferInfo is not None and TransferInfo['TransferStatus'] == 'done': # FIXME 'pennding' -> 'done'
+            wflow['TransferDone'] = True
+        return wflow
+
+    def getRucioRules(self, wflow, gran, rucioAcct):
+        """
+        Queries Rucio and builds the relevant list of blocklevel rules for
+        the given workflow
+        :param  wflow:   A MSRuleCleaner workflow representation
+        :param  gran:    Data granularity to search for Rucio rules. Possible values:
+                        'block' || 'container'
+        :return:        wflow
+        """
+        currPline = wflow['PlineMarkers'][-1]
+        # Find all the output placement rules created by the agents
+        # NOTE: Check MSOutput 'TransferStatus' first and assure it is 'done'
+        #       only then we proceed with building the respective list of rules
+        if wflow['TransferDone']:
+            for dataCont in wflow['OutputDatasets']:
+                if gran == 'container':
+                    wflow['RulesToClean'][currPline].extend(self.rucio.listDataRules(dataCont, account=rucioAcct))
+                elif gran == 'block':
+                    blocks = self.rucio.getBlocksInContainer(dataCont)
+                    for block in blocks:
+                        wflow['RulesToClean'][currPline].extend(self.rucio.listDataRules(block, account=rucioAcct))
+        return wflow
+
+    def cleanRucioRules(self, wflow):
+        """
+        Cleans all the Rules present in the field 'RulesToClean' in the MSRuleCleaner
+        workflow representation. And fills the relevant Cleanup Status.
+        :param wflow:   A MSRuleCleaner workflow representation
+        :return:        wflow
+        """
+        # NOTE: The function should be called independently and sequentially from
+        #       The Input and the respective BlockLevel pipelines.
+
+        # NOTE: The current functional pipeline is always the last one in the PlineMarkers list
+        currPline = wflow['PlineMarkers'][-1]
+        delResutls = []
+        if wflow['TransferDone']:
```

Comment:
Same comment as made on the function above. Some workflows will have `TransferDone=False`, thus not triggering rule deletion.

Reply:
Changed - The `TransferDone` flag is not checked  in the `cleanRucioRules()` method any more. We are about to guaranty only the proper workflows enter a cleaning pipeline.